x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/FrozenEngine.java

```diff
@@ -0,0 +1,475 @@
+            ensureOpenOrReset();
+            return delegate.getIndexCommit();
+        }
+
+        @Override
+        protected void doClose() throws IOException {
+            release();
+        }
+
+        @Override
+        public CacheHelper getReaderCacheHelper() {
+            ensureOpenOrReset();
+            return delegate.getReaderCacheHelper();
+        }
+
+        @Override
+        public DirectoryReader getDelegate() {
+            ensureOpenOrReset();
+            return delegate;
+        }
+    }
+
+    /**
+     * We basically duplicate a FilterLeafReader here since we don't want the
+     * incoming reader to register with this reader as a parent reader. This would mean we barf if the incoming
+     * reader is closed and that is what we actually doing on purpose.
+     */
+    static final class LazyLeafReader extends FilterLeafReader {
+
+        private volatile LeafReader in;
+        private final SegmentCommitInfo info;
+        private final int numDocs;
+        private final int maxDocs;
+
+        private LazyLeafReader(LeafReader in) {
+            super(Lucene.emptyReader(in.maxDoc())); // empty reader here to make FilterLeafReader happy
+            this.info = Lucene.segmentReader(in).getSegmentInfo();
+            this.in = in;
+            numDocs = in.numDocs();
+            maxDocs = in.maxDoc();
+            // don't register in reader as a subreader here.
+        }
+
+        private void ensureOpenOrReleased() {
+            ensureOpen();
+            if (in == null) {
+                throw new AlreadyClosedException("leaf is already released");
+            }
+        }
+
+        @Override
+        public Bits getLiveDocs() {
+            ensureOpenOrReleased();
+            return in.getLiveDocs();
+        }
+
+        @Override
+        public FieldInfos getFieldInfos() {
+            ensureOpenOrReleased();
+            return in.getFieldInfos();
+        }
+
+        @Override
+        public PointValues getPointValues(String field) throws IOException {
+            ensureOpenOrReleased();
+            return in.getPointValues(field);
+        }
+
+        @Override
+        public Fields getTermVectors(int docID)
+            throws IOException {
+            ensureOpenOrReleased();
+            return in.getTermVectors(docID);
+        }
+
+        @Override
+        public int numDocs() {
+            return numDocs;
+        }
+
+        @Override
+        public int maxDoc() {
+            return maxDocs;
+        }
+
+        @Override
+        public void document(int docID, StoredFieldVisitor visitor) throws IOException {
+            ensureOpenOrReleased();
+            in.document(docID, visitor);
+        }
+
+        @Override
+        protected void doClose() throws IOException {
+            in.close();
+        }
+
+        @Override
+        public CacheHelper getReaderCacheHelper() {
+            ensureOpenOrReleased();
```

Comment:
I'm not sure we want to delegate these, given that the point of the class is to allow the inner Reader to close, so we don't want anything to be cached on it?

Reply:
I think we need to do a lot more stuff to support that we have assertions all over the place that his is never null. In-fact its fine to cache this actually we ensure the lifetime of the reader is good enough?!